src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/logger.h"
+
+#include <seastar/core/loop.hh>
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size()) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                co_await item.client->stop();
```

Comment:
what prevents two fibers from calling `fetch_or_insert` on the same user, reaching this position in the code. one fiber calls stop(), the second fiber gets a reference to the same client and also calls stop. at that point the second call to stop will throw an exception because the clietn's gate is closed.

now that this method contains block co_await calls we need to think about concurrency.

there are a couple ways to do that. you could introduce a mutex. or another technique would be to stop the client in the background under a gate.

Reply:
Got a lesson from teacher @dotnwat about co-routines and their relationship to seastar fibers. A mutex may not be needed since the inclusion of eviction list.